_posts/2017-06-09-upgrade-to-rails-3.markdown

```diff
@@ -0,0 +1,157 @@
+
+- There is a new syntax for presence validations:
+
+```
+# Before:
+validates_presence_of :email
+
+# Now:
+validates :email, presence: true
+```
+
+- `ActiveRecord::Base.colorize_logging` and `config.active_record.colorize_logging` are deprecated in favor of `Rails::LogSubscriber.colorize_logging` or `config.colorize_logging`.
+
+### Action Mailer
+- `:charset`, `:content_type`, `:mime_version`, `:implicit_parts_order` are all deprecated in favor of `ActionMailer.default :key => value` style declarations.
+- Mailer dynamic `create_method_name` and `deliver_method_name` are deprecated, just call `method_name` which now returns a `Mail::Message` object.
+
+```
+# Before:
+message = UserMailer.create_welcome_email(user)
+UserMailer.deliver(message)
+
+# Now:
+message = UserMailer.welcome_email(user)
+message.deliver
+
+####
+
+# Before:
+UserMailer.deliver_welcome_email(user)
+
+# Now:
+UserMailer.welcome_email(user).deliver
+```
+- `template_root` is deprecated, pass options to a render call inside a proc from the `format.mime_type` method inside the mail generation block.
+- The body method to define instance variables is deprecated (`body {:ivar => value}`), just declare instance variables in the method directly and they will be available in the view.
+
+```
+# Before:
+def welcome_email(user)
+  ...
+  body {:user => user, :url => "https://fastruby.io"}
+end
+
+# Now:
+def welcome_email(user)
+  ...
+  @user = user
+  @url = "https://fastruby.io"
+end
+```
+- Mailers should now be in app/mailers instead of app/models.
+
+### Metal
+Since Rails 3 is closer to [Rack](http://guides.rubyonrails.org/rails_on_rack.html), the [Metal](http://weblog.rubyonrails.org/2008/12/17/introducing-rails-metal/) abstraction is no longer needed.
+
+This is the [official explanation](https://github.com/rails/rails/commit/ed34652d1aca148fea61c5309c1bd5ff3a55abfa) of what you need to do to update your existing Metals:
+
+- If your metal behaves like a middleware, add it to the middleware stack via config.middleware.use. You can use methods on the middleware stack to control exactly whereit should go.
+- If it behaves like a Rack endpoint, you can link to it in the router. This will result in more optimal routing time, and allows you to remove code in your endpoint that matches specific URLs in favor of the more powerful handling in the router itself.
+
+For the future, you can use ActionController::Metal to get a very fast controller with the ability to opt-in to specific controller features without paying the penalty of the full controller stack.
+
+### Railties
+[Railties](http://api.rubyonrails.org/classes/Rails/Railtie.html) deprecates the following constants during this version:
+
+- `RAILS_ROOT` in favor of `Rails.root`
+- `RAILS_ENV` in favor of `Rails.env`
+- `RAILS_DEFAULT_LOGGER` in favor of `Rails.logger`
+
+Also, PLUGIN/rails/tasks and PLUGIN/tasks are no longer loaded all tasks, now must be in PLUGIN/lib/tasks.
```

Comment:
@lubc Could you give an example about this? Also, please mark the paths as code, for example: `PLUGIN/rails/tasks`